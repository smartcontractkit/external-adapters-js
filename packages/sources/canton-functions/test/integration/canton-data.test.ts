import {
  TestAdapter,
  setEnvVariables,
} from '@chainlink/external-adapter-framework/util/testing-utils'
import * as nock from 'nock'
import * as process from 'process'
import {
  mockCantonApiExerciseChoiceOnLatestContractResponse,
  mockCantonApiExerciseChoiceResponse,
  mockCantonApiExerciseChoiceWithArgumentResponse,
  mockCantonApiQueryWithFilterResponse,
} from './fixtures'

describe('execute', () => {
  let spy: jest.SpyInstance
  let testAdapter: TestAdapter
  let oldEnv: NodeJS.ProcessEnv

  beforeAll(async () => {
    oldEnv = JSON.parse(JSON.stringify(process.env))
    process.env['AUTH_TOKEN'] = 'test-jwt-token'
    process.env['BACKGROUND_EXECUTE_MS'] = '1000'

    const mockDate = new Date('2025-10-14T00:00:00.000Z')
    spy = jest.spyOn(Date, 'now').mockReturnValue(mockDate.getTime())

    const adapter = (await import('../../src')).adapter
    adapter.rateLimiting = undefined
    testAdapter = await TestAdapter.startWithMockedCache(adapter, {
      testAdapter: {} as TestAdapter<never>,
    })
  })

  afterAll(async () => {
    setEnvVariables(oldEnv)
    await testAdapter.api.close()
    nock.restore()
    nock.cleanAll()
    spy.mockRestore()
  })

  describe('canton-data endpoint', () => {
    it('should return success when exercising a choice without argument', async () => {
      const data = {
        endpoint: 'canton-data',
        url: 'http://localhost:7575',
        templateId: 'example-package-id:Main:Asset',
        contractId: '00e1f5c6d8b9a7f4e3c2d1a0b9c8d7e6f5a4b3c2d1e0f9a8b7c6d5e4f3a2b1c0',
        choice: 'GetValue',
      }

      mockCantonApiExerciseChoiceResponse()

      const response = await testAdapter.request(data)
      expect(response.statusCode).toBe(200)
      expect(response.json()).toMatchSnapshot()
    }, 10000)

    it('should return success when exercising a choice with argument', async () => {
      const data = {
        endpoint: 'canton-data',
        url: 'http://localhost:7575',
        templateId: 'example-package-id:Main:Asset',
        contractId: '00e1f5c6d8b9a7f4e3c2d1a0b9c8d7e6f5a4b3c2d1e0f9a8b7c6d5e4f3a2b1c0',
        choice: 'UpdateValue',
        argument: JSON.stringify({ newValue: 2000 }),
      }

      mockCantonApiExerciseChoiceWithArgumentResponse()

      const response = await testAdapter.request(data)
      expect(response.statusCode).toBe(200)
      expect(response.json()).toMatchSnapshot()
    }, 10000)

    it('should query with filter and exercise choice on latest contract', async () => {
      const data = {
        endpoint: 'canton-data',
        url: 'http://localhost:7575',
        templateId: 'example-package-id:Main:Asset',
        contractFilter: JSON.stringify({ owner: 'Bob' }),
        choice: 'GetValue',
      }

      mockCantonApiQueryWithFilterResponse()
      mockCantonApiExerciseChoiceOnLatestContractResponse()

      const response = await testAdapter.request(data)
      expect(response.statusCode).toBe(200)
      expect(response.json()).toMatchSnapshot()
    }, 10000)
  })
})
