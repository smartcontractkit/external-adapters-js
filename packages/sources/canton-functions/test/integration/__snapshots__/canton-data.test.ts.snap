// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`execute canton-data endpoint should query with filter and exercise choice on latest contract 1`] = `
{
  "data": {
    "contract": {
      "agreementText": "Asset transfer agreement",
      "contractId": "33b4c8d9e1c2b0a7f6e5d4c3b2d1e0a9f8b7c6d5e4f3a2b1c0d9e8f7a6b5c4d3",
      "createdAt": "2025-10-17T12:00:00Z",
      "observers": [
        "Bob",
      ],
      "payload": {
        "amount": "2000",
        "currency": "USD",
        "isin": "US0378331005",
        "issuer": "Alice",
        "owner": "Bob",
      },
      "signatories": [
        "Alice",
      ],
      "templateId": "example-package-id:Main:Asset",
    },
    "exerciseResult": {
      "events": [
        {
          "contractId": "00e1f5c6d8b9a7f4e3c2d1a0b9c8d7e6f5a4b3c2d1e0f9a8b7c6d5e4f3a2b1c0",
          "eventId": "event-123",
          "eventType": "ChoiceExercised",
          "templateId": "example-package-id:Main:Asset",
        },
      ],
      "exerciseResult": {
        "currency": "USD",
        "value": "1000",
      },
      "status": 200,
    },
    "result": "{"exerciseResult":{"value":"1000","currency":"USD"},"events":[{"eventId":"event-123","contractId":"00e1f5c6d8b9a7f4e3c2d1a0b9c8d7e6f5a4b3c2d1e0f9a8b7c6d5e4f3a2b1c0","templateId":"example-package-id:Main:Asset","eventType":"ChoiceExercised"}],"status":200}",
  },
  "result": "{"exerciseResult":{"value":"1000","currency":"USD"},"events":[{"eventId":"event-123","contractId":"00e1f5c6d8b9a7f4e3c2d1a0b9c8d7e6f5a4b3c2d1e0f9a8b7c6d5e4f3a2b1c0","templateId":"example-package-id:Main:Asset","eventType":"ChoiceExercised"}],"status":200}",
  "statusCode": 200,
  "timestamps": {
    "providerDataReceivedUnixMs": 1760400000000,
    "providerDataRequestedUnixMs": 1760400000000,
  },
}
`;

exports[`execute canton-data endpoint should return success when exercising a choice with argument 1`] = `
{
  "data": {
    "exerciseResult": {
      "events": [
        {
          "contractId": "00e1f5c6d8b9a7f4e3c2d1a0b9c8d7e6f5a4b3c2d1e0f9a8b7c6d5e4f3a2b1c0",
          "eventId": "event-123",
          "eventType": "ChoiceExercised",
          "templateId": "example-package-id:Main:Asset",
        },
      ],
      "exerciseResult": {
        "currency": "USD",
        "value": "1000",
      },
      "status": 200,
    },
    "result": "{"exerciseResult":{"value":"1000","currency":"USD"},"events":[{"eventId":"event-123","contractId":"00e1f5c6d8b9a7f4e3c2d1a0b9c8d7e6f5a4b3c2d1e0f9a8b7c6d5e4f3a2b1c0","templateId":"example-package-id:Main:Asset","eventType":"ChoiceExercised"}],"status":200}",
  },
  "result": "{"exerciseResult":{"value":"1000","currency":"USD"},"events":[{"eventId":"event-123","contractId":"00e1f5c6d8b9a7f4e3c2d1a0b9c8d7e6f5a4b3c2d1e0f9a8b7c6d5e4f3a2b1c0","templateId":"example-package-id:Main:Asset","eventType":"ChoiceExercised"}],"status":200}",
  "statusCode": 200,
  "timestamps": {
    "providerDataReceivedUnixMs": 1760400000000,
    "providerDataRequestedUnixMs": 1760400000000,
  },
}
`;

exports[`execute canton-data endpoint should return success when exercising a choice without argument 1`] = `
{
  "data": {
    "exerciseResult": {
      "events": [
        {
          "contractId": "00e1f5c6d8b9a7f4e3c2d1a0b9c8d7e6f5a4b3c2d1e0f9a8b7c6d5e4f3a2b1c0",
          "eventId": "event-123",
          "eventType": "ChoiceExercised",
          "templateId": "example-package-id:Main:Asset",
        },
      ],
      "exerciseResult": {
        "currency": "USD",
        "value": "1000",
      },
      "status": 200,
    },
    "result": "{"exerciseResult":{"value":"1000","currency":"USD"},"events":[{"eventId":"event-123","contractId":"00e1f5c6d8b9a7f4e3c2d1a0b9c8d7e6f5a4b3c2d1e0f9a8b7c6d5e4f3a2b1c0","templateId":"example-package-id:Main:Asset","eventType":"ChoiceExercised"}],"status":200}",
  },
  "result": "{"exerciseResult":{"value":"1000","currency":"USD"},"events":[{"eventId":"event-123","contractId":"00e1f5c6d8b9a7f4e3c2d1a0b9c8d7e6f5a4b3c2d1e0f9a8b7c6d5e4f3a2b1c0","templateId":"example-package-id:Main:Asset","eventType":"ChoiceExercised"}],"status":200}",
  "statusCode": 200,
  "timestamps": {
    "providerDataReceivedUnixMs": 1760400000000,
    "providerDataRequestedUnixMs": 1760400000000,
  },
}
`;
